source/Nevermore/Serialization/InheritedClassConverter.cs

```diff
@@ -0,0 +1,130 @@
+
+            DocumentMap map = null;
+            RelationalMappings?.TryGet(value.GetType(), out map);
+
+            var documentType = value.GetType().GetTypeInfo();
+
+            foreach (var property in documentType
+                .GetProperties(BindingFlags.Instance | BindingFlags.Public | BindingFlags.GetProperty)
+                .Where(p => p.Name == TypeDesignatingPropertyName || 
+                            (p.CanRead && (map == null || (p.Name != map.IdColumn.Property.Name && map.IndexedColumns.All(c => p.Name != c.Property.Name))))))
+            {
+                writer.WritePropertyName(property.Name);
+                serializer.Serialize(writer, property.GetValue(value, null));
+            }
+
+            WriteTypeProperty(writer, value, serializer);
+
+            writer.WriteEndObject();
+        }
+
+        protected virtual void WriteTypeProperty(JsonWriter writer, object value, JsonSerializer serializer)
+        { }
+
+        protected virtual Type DefaultType { get; } = null;
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            if (reader.TokenType == JsonToken.Null)
+                return null;
+
+            var jo = JObject.Load(reader);
+            var designatingProperty = jo.GetValue(TypeDesignatingPropertyName);
+            TypeInfo typeInfo;
+            if (designatingProperty == null)
+            {
+                if (DefaultType == null)
+                {
+                    throw new Exception($"Unable to determine type to deserialize. Missing property `{TypeDesignatingPropertyName}`");
+                }
+                typeInfo = DefaultType.GetTypeInfo();
+            }
+            else
+            {
+                var derivedType = designatingProperty.ToObject<string>();
+                typeInfo = GetTypeInfoFromDerivedType(derivedType);
+            }
+
+            var ctor = typeInfo.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            if (ctor == null)
+            {
+                throw new Exception($"Type {typeInfo.Name} must have a public constructor");
+            }
+
+            var args = ctor.GetParameters().Select(p =>
+                jo.GetValue(char.ToUpper(p.Name[0]) + p.Name.Substring(1))?.ToObject(p.ParameterType, serializer)).ToArray();
+            var instance = ctor.Invoke(args);
+            foreach (var prop in typeInfo
+                .GetProperties(BindingFlags.Public | BindingFlags.SetProperty | BindingFlags.Instance)
+                .Where(p => p.CanWrite))
+            {
+                var val = jo.GetValue(prop.Name);
+                if (val != null)
+                    prop.SetValue(instance, val.ToObject(prop.PropertyType, serializer), null);
+            }
+            return instance;
+        }
+
+        protected virtual TypeInfo GetTypeInfoFromDerivedType(string derivedType)
```

Comment:
This confused me a bit, perhaps it would be clearer if you had a seperate class for the enum case and left this abstract. I think then you can also do `where TDerived : Enum`

Reply:
👍